doc/api/async_hooks.md

```diff
@@ -712,6 +712,75 @@ never be called.
 * Returns: {number} The same `triggerAsyncId` that is passed to the
 `AsyncResource` constructor.
 
+## Class: AsyncContext
+
+This class is used to create contexts that can be used through the event loop.
+
+### Example
+
+Let's build a logger that will always know the current HTTP request and use
+it to display enhanced logs without needing to explicitly pass the current
+HTTP request to it.
+
+```js
+const { AsyncContext } = require('async_storage');
+const http = require('http');
+
+const kReq = 'CURRENT_REQUEST';
+const asyncContext = new AsyncContext();
+
+function log(...args) {
+  const store = asyncContext.getStore();
+  // Let's make sure the store exists and it contains a request
+  if (store && store.has(kReq)) {
+    const req = store.get(kReq);
+    // For instance, prints `GET /items ERR could not do something
+    console.log(req.method, req.url, ...args);
+  } else {
+    console.log(...args);
+  }
+}
+
+http.createServer((request, response) => {
+  const store = asyncContext.enter();
+  store.set(kReq, request);
+  // some code
+   someAsyncOperation((err, result) => {
+    if (err){
+      log('ERR', err.message);
+      // ... rest of the code
+    }
+  });
+})
+.listen(8080);
+```
+
+### new AsyncContext()
+
+Creates a new instance of AsyncContext. Until the `enter` method is called, it
+does not provide any storage features.
+
+### asyncContext.enter()
+
+Calling `asyncContext.enter()` will create a new asynchronous context.
+This method returns a `Map` known as the store.
+
+This store will be persistent through the following asynchronous calls.
+
+### asyncContext.exit()
+
```

Comment:
I'll agree with some of the other comments that I'm not sure how/when I should call exit or what it will do for me (I'll admit I did not look at the implementation, only move some code using continuous-local-storage over to use these new apis)

Reply:
LGTM, let's remove `exit` ^^

Reply:
Can one do this just by calling `.create` and get an empty map? 

How does composing contexts work here?

Reply:
@benjamingr I am not a big fan of exit personally. I see a use case for it and it's probably clearer than starting a new trace with an empty map.

Contexts are independent from each others.

Reply:
This thing (async context) is implemented in other languages and libraries as well right? What does C# do? What do zones do in JavaScript? What does CLS do?

Reply:
Ah, the independence is achieved by having to hold an actual asyncStorage instance in order to get access to the data. Makes sense, this is different from other systems but I kind of like it.

Reply:
This is mainly a ext step from the CLS and the diverse implementation found in JS in APMs code.
 
C# has a simillar API named `AsyncLocal` (based on my understanding of a discussion with a cvoworker tbh.

Reply:
Yes, accessing the `asyncStorage` is on the user's responsibility. Which doesn't prevent anyone from attaching it to `process` and rebuild the CLS over this.